src/Week4/Four.hs

```diff
@@ -0,0 +1,76 @@
+module Week4.Four where
+
+import Data.Bits as Bits
+import Data.List
+
+---------- ex 1 ----------
+
+fun1' :: [Integer] -> Integer
+fun1' = product . map (subtract 2) . filter even
+
+fun2' :: Integer -> Integer
+fun2' = sum . filter even . takeWhile (>1) . iterate (\n -> if (even n) then (div n 2) else (3 * n + 1))
+
+---------- ex 2 ----------
+
+data Tree a = Leaf
+            | Node Integer (Tree a) a (Tree a)
+     deriving (Show, Eq)
+
+foldTree :: [a] -> Tree a
+foldTree = foldr insertBalanced Leaf
+    where
+        height :: Tree a -> Integer
+        height Leaf           = 0
+        height (Node h _ _ _) = h
+
+        insertBalanced :: a -> Tree a -> Tree a
+        insertBalanced e Leaf = Node 0 Leaf e Leaf
+        insertBalanced e t@(Node _ l _ r) =
+            case compare (height l) (height r) of
+                LT -> instertLeft e t
+                GT -> insertRight e t
+                EQ ->
+                    let inLeft = instertLeft e t
+                        inRight = insertRight e t
+                    in case compare (height inLeft) (height inRight) of
+                        GT -> inRight
+                        _  -> inLeft
+
+        instertLeft :: a -> Tree a -> Tree a
+        instertLeft e t@(Node h l x r) =
+            let addedLeft = insertBalanced e l
+                newHeight = (+1) $ max (height r) (height addedLeft)
+            in Node newHeight addedLeft x r
+
+        insertRight :: a -> Tree a -> Tree a
+        insertRight e t@(Node h l x r) =
+            let addedRight = insertBalanced e r
+                newHeight = (+1) $ max (height l) (height addedRight)
+            in Node newHeight l x addedRight
+
+
+---------- ex 3.1 ----------
+
+xor' :: [Bool] -> Bool
+xor' = foldr Bits.xor False
+
+---------- ex 3.2 ----------
+
+map' :: (a -> b) -> [a] -> [b]
+map' f = foldr (\a -> (++) [f a] ) []
```

Comment:
It's better to avoid appending of single-element list via `++`. Operator `:` is more efficient. Think how you can rewrite this lambda function to use `:` instead of `++`.